AADRoleEligibilityScheduleRequest - BUG - Subsequent Run Fails due to RoleAssignmentExists

### Description of the issue

This issue is similiar to https://github.com/microsoft/Microsoft365DSC/issues/5312. DSC builds during the first run and assigns the custom role to `sg-test-serviceprincipal-core`. However, during a subsequent run, the build fails due to the RoleAssignmentExists. I've tested this with built-in roles and that builds without issue. I believe this is due to the `Get-MgBetaRoleManagementDirectoryRoleEligibilityScheduleRequest` not return custom roles.
https://github.com/microsoft/Microsoft365DSC/blob/Dev/Modules/Microsoft365DSC/DSCResources/MSFT_AADRoleEligibilityScheduleRequest/MSFT_AADRoleEligibilityScheduleRequest.psm1#L188


Created Role:
![Image](https://github.com/user-attachments/assets/64604f53-b817-4793-a66c-c7b4bcbb45f4)

Security Group:
![Image](https://github.com/user-attachments/assets/20d1c9e2-dfc5-41ee-90a3-f4c1179a390a)




### Microsoft 365 DSC Version

1.24.1016.1

### Which workloads are affected

Azure Active Directory (Entra ID)

### The DSC configuration

#### Custom Role Definition

```
AADRoleDefinition 'AADRoleDefinition-CustomeRoleSecurityAnalyst'{
    DependsOn             = @(
        '[AADApplication]AADApplication-Microsoft365DSC'
    )
        DisplayName                   = '[Custom Role] Security Analyst'
        Description                   = 'Custom role'
        ResourceScopes                = '/'
        IsEnabled                     = $true
        RolePermissions               = @(
            'microsoft.directory/auditLogs/allProperties/read'
            'microsoft.directory/provisioningLogs/allProperties/read'
            'microsoft.directory/signInReports/allProperties/read'
            'microsoft.directory/auditLogs/allProperties/read'
            'microsoft.directory/deviceLocalCredentials/standard/read'
            'microsoft.directory/provisioningLogs/allProperties/read'
            'microsoft.directory/signInReports/allProperties/read'
        )
        Version                       = '1.0'
        Ensure                        = 'Present'
        ApplicationId                 = $ApplicationId
        TenantId                      = $TenantId
        CertificateThumbprint         = $Thumbprint
}
```

#### Group Definition
```
AADGroup 'sg-test-serviceprincipal-core' {
    DependsOn             = @(
        '[AADApplication]AADApplication-Microsoft365DSC'
    )
    DisplayName           = 'sg-test-serviceprincipal-core'
    MailNickname          = 'sg-test-serviceprincipal-core'
    Description           = 'sg-test-serviceprincipal-core test assignment of service principals' 
    # Id = ''
    Owners                = @()
    Members               = @('app-amsec-prod-basinpuller')
    # MemberOf = @()
    # GroupTypes            = @()
    # MembershipRule = ''
    # MembershipRuleProcessingState = ''
    SecurityEnabled       = $true 
    MailEnabled           = $false 
    IsAssignableToRole    = $true
    AssignedToRole        = @(
         ### L2|Active assignment permitted by RBAC team for this group.
    )
    Visibility            = 'Private'
    # AssignedLicenses =
    Ensure                = 'Present'
    ApplicationId         = $ApplicationId
    TenantId              = $TenantId
    CertificateThumbprint = $Thumbprint
}
```

#### Eligibility Schedule Request
```
    # https://microsoft365dsc.com/resources/azure-ad/AADRoleEligibilityScheduleRequest/
    AADRoleEligibilityScheduleRequest 'sg-test-serviceprincipal-core2' {
        DependsOn             = @(
            '[AADGroup]sg-test-serviceprincipal-core'
        )
        Principal             = 'sg-test-serviceprincipal-core' 
        RoleDefinition        = '[Custom Role] Security Analyst' 
        PrincipalType         = 'Group' 
        DirectoryScopeId      = "/"; 
        # Id
        # AppScopeId
        Action                = 'AdminAssign' 
        # IsValidationOnly
        Justification         = 'Assigning permanent eligibility' 
        IsValidationOnly      = $false
        ScheduleInfo          = MSFT_AADRoleEligibilityScheduleRequestSchedule {
            startDateTime = '2023-09-01T02:40:44Z'
            expiration    = MSFT_AADRoleEligibilityScheduleRequestScheduleExpiration {
                type = 'noExpiration'
            }
        }
        # TicketInfo
        Ensure                = 'Present'
        ApplicationId         = $ApplicationId
        TenantId              = $TenantId
        CertificateThumbprint = $Thumbprint
    }
```

### Verbose logs showing the problem

```
##[error][RoleAssignmentExists] : The Role assignment already exists.
    + CategoryInfo          : InvalidOperation: ({ Headers = , b...heduleRequest }:) [], CimException
    + FullyQualifiedErrorId : RoleAssignmentExists,Microsoft.Graph.Beta.PowerShell.Cmdlets.NewMgBetaRoleManagementDire 
   ctoryRoleEligibilityScheduleRequest_CreateExpanded
    + PSComputerName        : localhost
VERBOSE: [LCM-USWEST2]: LCM:  [ End    Set      ]  
[[AADRoleEligibilityScheduleRequest]sg-test-serviceprincipal-core2::[AzureAD]AzureAD_Configuration]  in 2.4390 seconds.
##[error]The PowerShell DSC resource 
'[AADRoleEligibilityScheduleRequest]sg-test-serviceprincipal-core2::[AzureAD]AzureAD_Configuration' with SourceInfo 'C:
\AzurePipeline-Agent\_work\1\s\M365Config\0.0.1\DSCResources\AzureAD\AzureAD.schema.psm1::106::5::AADRoleEligibilitySch
eduleRequest' threw one or more non-terminating errors while running the Set-TargetResource functionality. These 
errors are logged to the ETW channel called Microsoft-Windows-DSC/Operational. Refer to this channel for more details.
    + CategoryInfo          : InvalidOperation: (:) [], CimException
    + FullyQualifiedErrorId : NonTerminatingErrorFromProvider
    + PSComputerName        : localhost
```

### Environment Information + PowerShell Version

```
OsName               : Microsoft Windows Server 2022 Datacenter
OsOperatingSystemSKU : DatacenterServerEdition
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 20348.1.amd64fre.fe_release.210507-1500
OsLanguage           : en-US
OsMuiLanguages       : {en-US}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -60,6 +60,9 @@
   * Initial release.
 * AADRemoteNetwork
   * Initial release.
+* AADRoleEligibilityScheduleRequest
+  * Fixes for Custom roles.
+    FIXES [#5330](https://github.com/microsoft/Microsoft365DSC/issues/5330)
 * AADRoleManagementPolicyRule
   * Initial release.
 * AADServicePrincipal
@@ -67,6 +70,8 @@
   * FIXES [#5312](https://github.com/microsoft/Microsoft365DSC/issues/5312)
 * AADSocialIdentityProvider
   * Fixed missing permissions in settings.json
+* AADUserFlowAttribute
+  * Initial Release
 * AADVerifiedIdAuthority
   * Initial release.
 * AADVerifiedIdAuthorityContract
@@ -90,8 +95,6 @@
 * EXOArcConfig
   * Fixed `Test-TargetResource` to correctly check property `ArcTrustedSealers`
     when it has an array
-* AADUserFlowAttribute
-  * Initial Release
 * EXOMailboxAuditBypassAssociation
   * Initial release.
 * EXOMailboxSettings

--- a/Modules/Microsoft365DSC/DSCResources/MSFT_AADRoleEligibilityScheduleRequest/MSFT_AADRoleEligibilityScheduleRequest.psm1
+++ b/Modules/Microsoft365DSC/DSCResources/MSFT_AADRoleEligibilityScheduleRequest/MSFT_AADRoleEligibilityScheduleRequest.psm1
@@ -181,10 +181,15 @@
                     return $nullResult
                 }
                 Write-Verbose -Message "Found Principal {$PrincipalId}"
-                $RoleDefinitionId = (Get-MgBetaRoleManagementDirectoryRoleDefinition -Filter "DisplayName eq '$RoleDefinition'").Id
-                Write-Verbose -Message "Found Role {$RoleDefinitionId}"
-
-                $schedule = Get-MgBetaRoleManagementDirectoryRoleEligibilitySchedule -Filter "PrincipalId eq '$PrincipalId' and RoleDefinitionId eq '$RoleDefinitionId'"
+                $schedulesForPrincipal = Get-MgBetaRoleManagementDirectoryRoleEligibilitySchedule -Filter "PrincipalId eq '$PrincipalId'"
+                foreach ($instance in $schedulesForPrincipal)
+                {
+                    $roleInfo = Get-MgBetaRoleManagementDirectoryRoleDefinition -UnifiedRoleDefinitionId $instance.RoleDefinitionId
+                    if ($roleInfo.DisplayName -eq $RoleDefinition)
+                    {
+                        $schedule = $instance
+                    }
+                }
                 [Array]$request = Get-MgBetaRoleManagementDirectoryRoleEligibilityScheduleRequest -Filter "PrincipalId eq '$PrincipalId' and RoleDefinitionId eq '$RoleDefinitionId'" | Sort-Object -Property CompletedDateTime -Descending
 `
                 if ($request.Length -gt 1)
@@ -581,7 +586,7 @@ function Set-TargetResource
     {
         Write-Verbose -Message "Creating a Role Eligibility Schedule Request for user {$Principal} and role {$RoleDefinition}"
         $ParametersOps.Remove("Id") | Out-Null
-        Write-Verbose -Message "Current Values: $(Convert-M365DscHashtableToString -Hashtable $ParametersOps)"
+        Write-Verbose -Message "Values: $(Convert-M365DscHashtableToString -Hashtable $ParametersOps)"
         New-MgBetaRoleManagementDirectoryRoleEligibilityScheduleRequest @ParametersOps
     }
     elseif ($Ensure -eq 'Present' -and $currentInstance.Ensure -eq 'Present')

